sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala

```diff
@@ -138,3 +138,29 @@ object SpecialDatetimeValues extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Validate whether the [[Offset]] is valid.
```

Comment:
Let's explain the reason. `Dataset` API eagerly analyzes the query, so a query plan may contain invalid `Offset` operators but it's not the final query plan that gets evaluated.